components/nimbus/src/lib.rs

```diff
@@ -210,25 +238,109 @@ impl NimbusClient {
-                let state = self.mutable_state.lock().unwrap();
                 let evolver = EnrollmentsEvolver::new(
                     &nimbus_id,
                     &state.available_randomization_units,
-                    &self.app_context,
+                    &state.targeting_attributes,
                 );
-                let events = evolver.evolve_enrollments_in_db(db, &mut writer, &new_experiments)?;
-                self.database_cache.commit_and_update(db, writer)?;
-                events
+                evolver.evolve_enrollments_in_db(db, &mut writer, &new_experiments)?
             }
-            // We don't need to writer.commit() here because we haven't done anything.
             None => vec![],
-        })
+        };
+        self.database_cache.commit_and_update(db, writer)?;
+        Ok(res)
+    }
+
+    fn get_installation_date(&self, db: &Database, writer: &mut Writer) -> Result<DateTime<Utc>> {
+        let store = db.get_store(StoreId::Meta);
+        let persisted_installation_date: Option<DateTime<Utc>> =
+            store.get(writer, DB_KEY_INSTALLATION_DATE)?;
+        let res = if let Some(installation_date) = persisted_installation_date {
+            installation_date
+        } else {
+            let installation_date = match self.get_creation_date_from_path(&self.db_path) {
+                Ok(installation_date) => installation_date,
+                Err(_) => Utc::now(),
+            };
+            let store = db.get_store(StoreId::Meta);
+            store.put(writer, DB_KEY_INSTALLATION_DATE, &installation_date)?;
+            installation_date
+        };
+        Ok(res)
+    }
+
+    fn get_update_date(&self, db: &Database, writer: &mut Writer) -> Result<DateTime<Utc>> {
+        let store = db.get_store(StoreId::Meta);
+
+        let persisted_app_version: Option<String> = store.get(writer, DB_KEY_APP_VERSION)?;
+        let res = match (persisted_app_version, &self.app_context.app_version) {
+            (persisted_app_version, Some(current_app_version))
+                if (persisted_app_version.is_some()
+                    && persisted_app_version.clone().unwrap() != *current_app_version)
+                    || persisted_app_version.is_none() =>
+            {
+                store.put(writer, DB_KEY_APP_VERSION, current_app_version)?;
+                let update_date = Utc::now();
+                store.put(writer, DB_KEY_UPDATE_DATE, &update_date)?;
+                update_date
+            }
+            _ => {
+                let update_date: Option<DateTime<Utc>> = store.get(writer, DB_KEY_UPDATE_DATE)?;
+                if let Some(update_date) = update_date {
+                    update_date
+                } else {
+                    let update_date = Utc::now();
+                    store.put(writer, DB_KEY_UPDATE_DATE, &update_date)?;
+                    update_date
+                }
+            }
+        };
```

Comment:
The logic in this method looks like it could be cleaned up. I don't think we should touch the database if there is no version in the `app_context`, so our match statement ends up as:

```rust
(Some(persisted), Some(update_date)) if persisted == current => update_date,
(_, _) => {
     // write the current and now to the datase
     // return now.
} 
```